### PR TITLE
Implement Offline Recording and Navigation with UI and Permission Handling

### PR DESCRIPTION
**Pull request by helkhour**
Description:

This PR focuses on implementing the offline recording functionality within the app, addressing the user story where users can record themselves online, listen to their recordings, and store them locally. The implementation is progressing step by step with thorough testing, ensuring a robust offline user experience.

Files Added and Modified:

    OfflineRecordingScreen.kt:
        Introduced OfflineRecordingScreen for offline recording functionality.
       I tried to animate the microphone a bit , hope you like it ! 
I also wanted to maintain the question for the user, but this meant adding a navigation logic in MainActivity. I'll try to fix this in future PRs but couldn't manage to find a solution this time ... 

    OfflinePracticeQuestionsScreen.kt:
        Updated to navigate dynamically to OfflineRecordingScreen with selected questions as parameters.
        I tried to enhance UI with improved navigation and styling using MaterialTheme.colorScheme.

I also  added a composable route for OfflineRecordingScreen with {question} parameter to facilitate dynamic navigation.

    Tests:
        Added tests to ensure proper functionality of navigation actions and permission handling.
        Integrated color testing from MaterialTheme.colorScheme to maintain UI consistency.

Future Enhancements (in #68):
    In future iterations (#68), plan to enable offline playback functionality, allowing users to listen to their recordings offline.

Your feedback and suggestions for further improvements are appreciated!

### CODE REVIEW

**Comment by helkhour** [comment]
SonarCloud is not passing because the line coverage of NavigationActions is 0%. All others are >80 % -> answering our requirements ...

**Review by Rolph1 — APPROVED** [review]
Great pull request that adds an important feature to the app!

**Things you did well on:**
I can see this offline feature took a lot of work and your code seems well modularized and clean! Not only that but you also managed to still follow the MVVM architecture pattern to maximize line coverage even though you didn't have to.

The libraries you used seem quite advanced and your code looks AI free so I'm sure you had to read a lot of documentation. I really appreciate all the effort you put in to make quality code and understand everything in depth.


**Things you could work on:**
You have way too many lines in a single pull request! The assistants have already told us that no pull request needs to be over 1k lines long. You should also have multiple commits that each add a new "aspect" of the feature.

You should add TODOs in the code itself where the new feature should be implemented so that we could work on them later, and add a new task in the scrum board for that.

You should work more on the documentation! The code you wrote seems pretty difficult to understand (and good job on figuring it out!) but I can't really tell what it does with just the comments you wrote.

In OfflinePracticeQuestions the padding is hardcoded, please fix this in your next commits by using constants Baptiste added in his last commit.

**In short:**
Great quality code that adds important features to our app and must have taken a lot of time. For future commits, please separate your work into different commits, and make sure to use constants instead of hardcoding paddings/spacers.

File: app/src/main/java/com/github/se/orator/ui/offline/OfflineRecordingScreen.kt [hunk 1]
```diff
@@ -4,93 +4,140 @@ import android.Manifest
...
+        permissionGranted = isGranted
+        if (isGranted) {
+          speakingViewModel.onMicButtonClicked(true)
+        }
+      }
 
-  // Request recording permission on initial composition
-  LaunchedEffect(Unit) { permissionLauncher.launch(Manifest.permission.RECORD_AUDIO) }
+  val isRecording by speakingViewModel.isRecording.collectAsState()
+  val infiniteTransition = rememberInfiniteTransition()
```
Comment (charbelraffoul) [inline app/src/main/java/com/github/se/orator/ui/offline/OfflineRecordingScreen.kt:53]: We should probably use the same Visualizer for the recording action for both online and offline mode. Here is the code snippet from the method we use in SpeakingScreen: 

"/**
 * A composable that visualizes audio amplitudes as a waveform.
 *
 * @param amplitudes A list of amplitude values to visualize.
 */
@Composable
fun AudioVisualizer(amplitudes: List<Float>) {
  Canvas(
      modifier =
          Modifier.fillMaxWidth()
              .height(AppDimensions.visualizerHeight)
              .testTag("audio_visualizer")) {
        val width = size.width
        val height = size.height
        val barWidth = width / amplitudes.size
        amplitudes.forEachIndexed { index, amplitude ->
          val barHeight = (amplitude / Short.MAX_VALUE) * height
          drawLine(
              color = AppColors.primaryColor,
              start = Offset(x = index * barWidth, y = height / 2 - barHeight / 2),
              end = Offset(x = index * barWidth, y = height / 2 + barHeight / 2),
              strokeWidth = barWidth)
        }
      }". This method make the visualization dependent on the user's voice, and adds UI consistency to our app!

File: app/src/main/java/com/github/se/orator/ui/offline/OfflineRecordingScreen.kt [hunk 2]
```diff
@@ -4,93 +4,140 @@ import android.Manifest
...
+  val scale by
+      infiniteTransition.animateFloat(
+          initialValue = 1f,
+          targetValue = 1.3f,
+          animationSpec =
+              infiniteRepeatable(
+                  animation = tween(600, easing = LinearEasing), repeatMode = RepeatMode.Reverse))
```
Comment (charbelraffoul) [inline app/src/main/java/com/github/se/orator/ui/offline/OfflineRecordingScreen.kt:60]: We should probably use the same Visualizer for the recording action for both online and offline mode. Here is the code snippet from the method we use in SpeakingScreen:

"/**

A composable that visualizes audio amplitudes as a waveform.
@param amplitudes A list of amplitude values to visualize.
*/
https://github.com/composable
fun AudioVisualizer(amplitudes: List) {
Canvas(
modifier =
Modifier.fillMaxWidth()
.height(AppDimensions.visualizerHeight)
.testTag("audio_visualizer")) {
val width = size.width
val height = size.height
val barWidth = width / amplitudes.size
amplitudes.forEachIndexed { index, amplitude ->
val barHeight = (amplitude / Short.MAX_VALUE) * height
drawLine(
color = AppColors.primaryColor,
start = Offset(x = index * barWidth, y = height / 2 - barHeight / 2),
end = Offset(x = index * barWidth, y = height / 2 + barHeight / 2),
strokeWidth = barWidth)
}
}". This method make the visualization dependent on the user's voice, and adds UI consistency to our app!

File: app/src/main/java/com/github/se/orator/ui/offline/OfflineRecordingScreen.kt [hunk 3]
```diff
@@ -4,93 +4,140 @@ import android.Manifest
...
 
-  // Layout for the offline speaking screen
   Column(
-      modifier = Modifier.fillMaxSize().padding(16.dp),
-      verticalArrangement = Arrangement.Center,
+      modifier =
+          Modifier.fillMaxSize()
+              .background(colors.background)
+              .padding(WindowInsets.systemBars.asPaddingValues())
+              .padding(horizontal = 16.dp)
```
Comment (charbelraffoul) [inline app/src/main/java/com/github/se/orator/ui/offline/OfflineRecordingScreen.kt:76]: Avoid Hardcoded UI Values: I noticed that some UI elements are using hardcoded padding and sizing values. To ensure a consistent and responsive design across different screen sizes, consider using the createAppDimensions function. This approach allows padding and dimensions to adapt based on the device's screen dimensions, enhancing the overall user experience by providing a more adaptable and visually appealing interface.

**Review by charbelraffoul — CHANGES_REQUESTED** [review]
# Pull Request Review: Implementing Offline Recording Functionality

Great work on implementing the offline recording functionality! 🎉

## **Highlights:**

### 1. `OfflineRecordingScreen.kt`
- **Microphone Animation:**  
  The animated microphone adds a dynamic and engaging touch to the recording experience. It enhances user interaction and makes the UI more lively. But I think it should be better to use the same animations from the SpeakingScreen of the online mode, which were updated with an AudioVisualizer.
  
- **Navigation Logic:**  
  Successfully maintaining the question context by passing it through navigation is a crucial step. I understand the challenges you faced with `MainActivity` navigation logic. Looking forward to seeing how you refine this in future PRs!

### 2. `OfflinePracticeQuestionsScreen.kt`

- **UI Enhancements:**  
  Utilizing `MaterialTheme.colorScheme` for styling ensures consistency with the app’s overall theme, providing a seamless user experience.

### 3. **Testing:**
- **Comprehensive Tests:**  
  Adding tests for navigation actions and permission handling significantly boosts the reliability of the new features. Ensuring UI consistency through color testing is also a great move.
  
- **Future Test Coverage:**  
  As you continue to build out offline functionalities, maintaining and expanding test coverage will be essential for sustained quality.

## **Suggestions:**

- **Avoid Hardcoded UI Values:**  
  I noticed that some UI elements are using hardcoded padding and sizing values. To ensure a consistent and responsive design across different screen sizes, consider using the `createAppDimensions` function. This approach allows padding and dimensions to adapt based on the device's screen dimensions, enhancing the overall user experience by providing a more adaptable and visually appealing interface.

- **User Feedback:**  
  Incorporate more user feedback mechanisms within the `OfflineRecordingScreen`, such as visual indicators for recording status or error messages for permission denials, to enhance usability.

- **Documentation:**  
  Adding more detailed comments or documentation within the new composables can help future developers understand the flow and purpose of each component more quickly.

## **Looking Ahead:**

Excited to see the planned offline playback functionality in #68! This will round out the offline experience and provide users with a complete set of tools for practicing and reviewing their recordings.

Overall, this PR lays a solid foundation for offline capabilities with thoughtful UI design and robust testing. Keep up the excellent work! I am

## **Requested Changes:**

- **Remove Hardcoded UI Values:**  
  Please remove the hardcoded padding and sizing values used in the UI components. Instead, utilize the `createAppDimensions` function to make the padding and dimensions responsive to different screen sizes. This change will enhance the user experience by ensuring the UI adapts smoothly across various devices.

- **Integrate Audio Visualizer:**  
  Replace the current microphone animation with the AudioVisualizer used in the online SpeakingScreen. This integration will maintain consistency across recording functionalities and provide users with a more informative and visually appealing recording experience.

---

*Feel free to reach out if you need any further feedback or assistance with upcoming features!*

**Comment by helkhour** [comment]
I created new issues #129 for the audio visualizer integration and #128 for the hardcoded UI values
